### DISCUSSION TITLE
Code completion for magic (double dunder) method signatures after typing `def __`

### DISCUSSION THREAD

**Discussion by maltevesper** [microsoft/pylance-release#4865] [Enhancement]
Could you please reconsider microsoft/vscode-python#602? (Completions for magic methods)

Everytime I write a context manager I have to log up the signature of the exit/aexit method, since there is no completion for the function signature.

In 602 it was stated that magic methods would clutter the completion suggestions. If the completions for magic methods would only be shown after I typed `def __`, they should not be a nuance, since at that point it should be pretty clear that I am trying to implement a magic method. 

In particullar this suggestion https://github.com/microsoft/vscode-python/issues/602#issuecomment-359701094 shows exactly what I am hoping for.

**Comment by rchiodo** [comment]
Thanks for the suggestion. I still think the same reasoning applies though. We can't list every possible magic method. That's why the completions don't work. 

For example, `__init__` works just fine. If you type a class and then do `def __in` and hit tab, it will complete it like so:

```python
class Foo():
    def __init__(self) -> None:
        pass # pass is highlighted so typing will just start replacing the word pass.
```

Maybe there are specific ones we could add though? Not sure how we came up with the list we currently support.

**Comment by erictraut** [comment]
Some magic methods in Python have well-defined signatures that could be hard-coded as code snippets in a tool like pylance. For example, every `__enter__` has the same signature every time it's used. Other methods like `__exit__` and `__iter__` have the same signature 90% of the time, so a hard-coded code snippet would probably work for those too.

Currently, pylance doesn't have any knowledge of specific magic methods. Its completions are based on overrides. If a base class (e.g. `object`) declares that method, then it will appear as a completion suggestion. For example, the `def __eq__` example cited in the linked issue works fine in pylance today because all classes have an implicit `object` base class, and `object` defines an `__eq__` method.

Most magic methods have an associated abstract class or protocol class defined in the `typing` or `builtins` module. For example, if you want to create a context manager, your class would typically derive from `typing.ContextManager`, which is an abstract base class for a context manager. If you add this as a base class, pylance will then auto-complete the overrides for `__enter__` and `__exit__`. 

@maltevesper, which magic methods do you use on a regular basis that are not auto-completed by pylance today?

**Comment by maltevesper** [comment]
Sorry about the late response, I had a little play and a good think, here are the points I cam up with:

From what I can tell in [PEP343](https://peps.python.org/pep-0343/) all `__enter__` and `__exit__` methods should have the same signatures, unless one accounts for leading underlines if all the exception info is ignored in the exit methods. Since I mostly dont want to remember the three exception info arguments and their order, this is where I would appreciate auto completion most.

Generally I think https://docs.python.org/3/reference/datamodel.html is a good overview of dunder methods with fixed signatures.

That said, I had a look at `contextlib.AbstractContextManager` (`typing.ContextManager`). While adding the base class did help with getting typing suggestion, I cant really see myself using this much. The abstract base class does not give me a real advantage (If I Implement context managers I tend to do more than the ABC's default implementation in enter and exit anyways).
That said [PEP544](https://peps.python.org/pep-0544/) hints, that there should be little need to inherit as well:
"[...]
The same problem appears with user-defined ABCs: they must be explicitly subclassed or registered. This is particularly difficult to do with library types as the type objects may be hidden deep in the implementation of the library. Also, extensive use of ABCs might impose additional runtime costs.

The intention of this PEP is to solve all these problems by allowing users to write the above code without explicit base classes in the class definition, allowing Bucket to be implicitly considered a subtype of both Sized and Iterable[int] by static type checkers using structural [[wiki-structural]](https://peps.python.org/pep-0544/#wiki-structural) subtyping:
[...]"

While I personally care most about code completions for methods with "complicated signatures" where the arguments are not obvious (i.e. exit/aexit), I still think completions for all methods would be great. That said, personally I find my self most often using (a)enter/(a)exit. I also use the functions to create custom mutable mappings, and would appreciate autocompletions there, but I find their parameters to be much more self explanitory.

On a different note, I noticed if I do inherit i.e. `collections.abc.MutableMapping` I do get the completions for most methods, but no warning about not implemented `@abstractmethod`s. I am not sure if that would be helpful or harmful. Also the autogeneration often generates code of the form:
```
def __something__(...) -> None:
    return super().__something__(...)
```
while technically correct, it looks funny to have a `return` in a function returning None. But that is purely cosmetic.

**Comment by erictraut** [comment]
> I do get the completions for most methods, but no warning about not implemented `@abstractmethod`.

You will see an error if you mark the class as `@final`, which means that no subclasses are possible. If the class is not marked `@final`, then it's possible that subclasses could define the missing abstract methods.

Also, you will see errors if you attempt to instantiate (i.e. call the constructor of) a class that doesn't implement all of its abstract methods. That's the point where an error would be detected and reported at runtime.

---

> From what I can tell in [PEP343](https://peps.python.org/pep-0343/) all __enter__ and __exit__ methods should have the same signatures.

The return type for both of these methods may vary, but the input parameters should always have the same type.

**Comment by maltevesper** [comment]
That makes sense, I just am spoiled by i.e. the dart extension which provides quick actions to add the missing methods. But given the explanation about subclassing my hunch about it being harmful seems confirmed.That said let's get back on topic from which I diverted: what would be the best way forward for good completions for dunder methods once the `def __` has been typed?

**Comment by debonte** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by PhenomenaPh** [comment]
Any updates on this issue? Coming from PyCharm, I find it very strange that basic completion for __ methods is missing. For example, PyCharm offers autocompletion for all standard magic methods, like this:

<img width="537" alt="Screenshot 2025-04-21 at 16 14 14" src="https://github.com/user-attachments/assets/ef558a0f-b80b-443b-a877-e7eac4d056ae" />
